\PassOptionsToPackage{dvipsnames,table}{xcolor}
\documentclass[final]{beamer}
\mode<presentation> {
  \usefonttheme[onlymath]{serif}
}
\usepackage[absolute,overlay]{textpos}
\usepackage{tikz}
\usetikzlibrary{shapes,arrows}
\usepackage[mathscr]{euscript}
\usepackage{amssymb}
\usepackage{tangocolors}
\usepackage{array}
\usepackage{tabularx}
\usepackage{color}
\usepackage{booktabs}
\setbeamercolor{headline}{fg=tabutter,bg=taaluminium}
\setbeamercolor{separation line}{bg=ta2orange}
\setbeamercolor{title in headline}{fg=ta3skyblue}
\setbeamercolor{author in headline}{fg=ta2skyblue}
\setbeamercolor{institute in headline}{fg=ta2skyblue}

\setbeamercolor{framesubtitle}{fg=ta3orange, bg=ta2gray}
\setbeamercolor{author in head/foot}{fg=ta2orange, bg=black}
\setbeamercolor{title in head/foot}{fg=ta2orange, bg=black}

\setbeamercolor*{normal text}{fg=tachameleon, bg=ta3gray}
\setbeamercolor*{block body}{bg=white,fg=ta3gray}
\setbeamercolor{upper separation line head}{fg=ta2orange}

\setbeamerfont{block title}{size=\Large}
\setbeamercolor*{block title}{fg=black!70!blue,bg=taaluminium}

\setbeamercolor*{block details title}{fg=taorange!20!yellow!70,bg=ta2gray}
\setbeamerfont{block details title}{size=\Large}

\setbeamercolor*{example body}{fg=ta3aluminium,bg=black}
\setbeamercolor*{example text}{fg=ta3aluminium,bg=black}
\setbeamercolor*{example title}{bg=taorange,fg=ta2gray}

\setbeamercolor{alerted text}{fg=ta3gray}

\setbeamercolor{structure}{fg=ta3skyblue}

\setbeamertemplate{navigation symbols}{}

\setbeamertemplate{block begin}{
  \vskip.75ex
  \begin{beamercolorbox}[rounded=true,shadow=true,leftskip=1cm,colsep*=.75ex]{block title}%
    \usebeamerfont*{block title}\insertblocktitle
  \end{beamercolorbox}%
  {\ifbeamercolorempty[bg]{block body}{}{\nointerlineskip\vskip-0.5pt}}%
  \usebeamerfont{block body}%
  \begin{beamercolorbox}[rounded=true,shadow=true,colsep*=.75ex,sep=.75ex,vmode]{block body}%
    \ifbeamercolorempty[bg]{block body}{\vskip-.25ex}{\vskip-.75ex}\vbox{}%
}

\setbeamertemplate{block end}{
  \end{beamercolorbox}
}

% Introducing new environment for blocks with details to distinguish them visually
\mode
<all>
{
  \newenvironment<>{detailsblock}[1]{%
    \begin{actionenv}#2%
      \def\insertblocktitle{#1}%
      \par%
      \usebeamertemplate{block details begin}}
    {\par%
      \usebeamertemplate{block details end}%
    \end{actionenv}}
}


\setbeamertemplate{block details begin}{
  \vskip.75ex
  \begin{beamercolorbox}[rounded=true,shadow=true,leftskip=1cm,colsep*=.75ex]{block details title}%
    \usebeamerfont*{block details title}\insertblocktitle
  \end{beamercolorbox}%
  {\ifbeamercolorempty[bg]{block body}{}{\nointerlineskip\vskip-0.5pt}}%
  \usebeamerfont{block body}%
  \begin{beamercolorbox}[rounded=true,shadow=true,colsep*=.75ex,sep=.75ex,vmode]{block body}%
    \ifbeamercolorempty[bg]{block body}{\vskip-.25ex}{\vskip-.75ex}\vbox{}%
}

\setbeamertemplate{block details end}{
  \end{beamercolorbox}
}

\setbeamertemplate{headline}{  
  \leavevmode

  \begin{beamercolorbox}[wd=\paperwidth]{headline}
%    \begin{columns}[T]
%      \begin{column}{.1\paperwidth}
%      \end{column}
%      \begin{column}{.8\paperwidth}
        \vskip4ex
        \centering
        \usebeamercolor{title in headline}{\color{fg}\textbf{\Huge{\inserttitle}}\\[1ex]}
        \usebeamercolor{author in headline}{\color{fg}\LARGE{\insertauthor}\\[1ex]}
        \usebeamercolor{institute in headline}{\color{fg}\large{\insertinstitute}\\[1ex]}     
%      \end{column}
%      \begin{column}{.1\paperwidth}
%      \end{column}
%    \end{columns}
  \end{beamercolorbox}

  \begin{beamercolorbox}[wd=\paperwidth]{lower separation line head}
    \rule{0pt}{2pt}
  \end{beamercolorbox}
}

\setbeamertemplate{footline}[page number]{}

\usepackage[utf8x]{inputenc}
\usepackage[english]{babel}
\usepackage{xcolor}
\usepackage{graphicx}

\usepackage[scale=1.0]{beamerposter}
\setlength{\paperwidth}{48in}
\setlength{\paperheight}{36in}
\usepackage[noend]{algpseudocode}
\usepackage{parskip}

\setbeamertemplate{caption}[numbered]
\setbeamertemplate{footline}{}

\newcommand{\columncaption}[1]{
\begin{center}
  {\huge #1}
\end{center}
}

\author{Anton Korobeynikov$^{1,2}$, Artem Tarasov$^{1}$}
\title{HammerIT: homopolymer-space Hamming clustering for IonTorrent read error correction}
\institute{$^{1}$ Department of Statistical Modelling, St. Petersburg State University, St. Petersburg, Russia;
           $^{2}$ Algorithmic Biology Laboratory, St. Petersburg Academic University, St. Petersburg, Russia
}

\setlength{\TPHorizModule}{\paperwidth}
\setlength{\TPVertModule}{\paperheight}
\begin{document}


\newcolumntype{C}[1]{>{\vbox to 1em\bgroup\vfill\centering}p{#1} <{\egroup}} 

\begin{frame}

    \begin{textblock}{0.450}(0.01, 0.075)
      \begin{block}{Introduction\phantom{p}}
        Error correction of sequenced reads remains a difficult task,
        especially for data obtained using IonTorrent technology due to its
        higher error rate. The task is even more challenging in single-cell
        sequencing projects with extremely non-uniform coverage.

        The existing error correction tools assume that the most sequencing
        errors in the data are mismatches and thus perform poorly on
        IonTorrent data with its prevailing errors due to homopolymer
        indels.

        We introduce \textsc{HammerIT}~--- a novel error correction tool
        which is specifically tuned for IonTorrent sequencing
        errors.
      \end{block}
    \end{textblock}

    \begin{textblock}{0.215}(0.01,0.190)
    \begin{block}{IonTorrent error profile}
      Corrected flow signal intensities are available in BAM files
      produced by versions of Ion Torrent Suite prior to 3.4. Called
      homopolymer length is obtained as corrected flow signal
      intensity rounded to the nearest integer.

      We have studied flow signal intensity distributions around
      insertion/deletion sites. File B7-295.bam, downloaded from Ion
      Community website, contained 4.6M~insertions, 5.0M~deletions,
      and 1.5M~mismatches.
      
      Overwhelming majority of errors turned out to be
      insertions/deletions of length~1, occuring when flow signal
      intensity is approximately halfway between two adjacent
      integers.
      \vspace{0.4em}

      \begin{figure}[h!]
        \caption{Flow signal intensities at insertion sites}
        \includegraphics[width=\textwidth]{images/overcalls}
      \end{figure}

      \begin{figure}[h!]
        \caption{Flow signal intensities at deletion sites}
        \includegraphics[width=\textwidth]{images/undercalls}
      \end{figure}
%      More detailed analysis of errors in IonTorrent data can be found
%      in the article
%      ``Shining a Light on Dark Sequencing: Characterising Errors in
%      Ion Torrent PGM Data'' \emph{(PLoS~Comput~Biol~9(4))}     
    \end{block}
  \end{textblock}


  \begin{textblock}{0.230}(0.230, 0.190)
    \begin{block}{Notation\phantom{p}}
      %{\large
      Let $\mathscr{N}$ denote the nucleotide alphabet $\left\{\textrm{A, C, G, T}\right\}$.

      By~definition, put $\mathscr{H} = \mathscr{N} \times \mathbb{N}^+$.

      We call an~element of~the~alphabet~$\mathscr{H}$~a~\emph{homopolymer run},
      and~an~element of~$\mathscr{H}^k$~a~\emph{homopolymer-space $k$-mer}.

      We use $x[k]$ to denote $k$-th element of
      a sequence, using zero-based indexing; $x[k\textrm{ .. }l]$
      to denote a subsequence $x[k]x[k+1]\dots x[l]$; finally, length of $x$ is denoted by $|x|$.
      
      Distance between $i,\;j \in \mathscr{H}^k$  is defined as the
      minimum number of $1$-base insertions/deletions/mismatches needed to align common part of $i$ and $j$.
      It is denoted by $\mathrm{dist}(i, j)$.
      %}
    \end{block}

    \begin{block}{HammerIT workflow}
      \vspace{-.1em}
      \begin{figure}
        \centering
        \tikzstyle{state} = [draw, ellipse, fill=green!20, text centered, minimum width=16em]
        \tikzstyle{block} = [draw, rectangle, rounded corners, fill=blue!10, text centered, minimum width=24em]
        \tikzstyle{line} = [draw, -triangle 45]
        \resizebox{\columnwidth}{!}{
        \begin{tikzpicture}[node distance = 2.5cm]
          \node[state] (raw) {Set of reads};
          \node[block, below of=raw] (count) {Count homopolymer-space $k$-mer statistics};
          \node[block, below of=count] (cluster) {Single-linkage clustering of homopolymer-space $k$-mers};
          \node[block, below of=cluster] (subcluster) {Subclustering with $k$-means};
          \node[block, below of=subcluster, text width=18em] (correction) {Correcting reads};
          \node[state, below of=correction] (corrected) {Set of corrected reads};
          \path[line] (raw) -- (count);
          \path[line] (count) -- (cluster);
          \path[line] (cluster) -- (subcluster);
          \path[line] (subcluster) -- (correction);
          \path[line] (correction) -- (corrected);
        \end{tikzpicture}
        }
      \end{figure}
    \end{block}
    
  \end{textblock}

  \begin{textblock}{0.320}(0.465, 0.075)
    \begin{block}{Error reduction results\phantom{p}}

      We evaluated HammerIT on 6 publicly available datasets,
      using the same pipeline as the authors of the recently published
      article \emph{(JÃ¼nemann~et~al,~Nat.~Biotech.,~2013;~vol.~31,~p.294--296)}.
      In that article, error rate in four Ion~Torrent datasets has
      been assessed. We used the same data plus two extra datasets from
      314v2~chip, which recently became available on Ion~Community~Portal.
     
      Indel/mismatch error rates were calculated for uniquely mapped
      reads before and after correction. For each dataset, correction
      was done in two ways. In the first setup, trimming was done for 
      read ends that couldn't be corrected due to lack of good k-mers, 
      while in the second one such read ends were preserved in the
      output. Relative change in read coverage after
      correction stayed within 0.4\% in all cases.

      \vspace{0.64em}

      \begin{figure}[h]
        \caption{Error rates before and after correction}
        \includegraphics[width=0.947\textwidth]{images/error_rate}
      \end{figure}

      \begin{figure}[h]
        \caption{Error reduction by read position for Sakai 400bp reads}
        \includegraphics[width=0.947\textwidth]{images/error_rate_vs_offset}
      \end{figure}
    \end{block}
  \end{textblock}
\begin{textblock}{0.205}(0.790, 0.075)
    \begin{block}{Assembly results}
      We have assembled \emph{E.coli O157 H7 Sakai} 400bp reads before and after correction with SPAdes 2.4.0.
      Read coverage is $_{\widetilde{~}}$240x, length of the reference genome is 5498450 bp.

      \vspace{0.1em}
      \begin{table}[ht]
        \caption{Assembly results. Contigs of length $\geq 500$ are used.}
      \begin{tabularx}{0.95\textwidth}{>{\fontsize{21.5}{24}\fontencoding{T1}\fontfamily{lmss}\fontseries{b}\selectfont}X|c|c}\toprule
                                                   & uncorrected & corrected \\ \midrule
        \# contigs                 \rule{0pt}{4ex} & 266         & 242       \\[1.0em]
        Largest contig                             & 316904      & 374932    \\[1.0em]
        Total length                               & 5322223     & 5320566   \\[1.0em]
        NG50                                       & 106242      & 146551    \\[1.0em]
        NG75                                       & 41186       & 44318     \\[1.0em]
        \# misassemblies                           & 0           & 2         \\[1.0em]
        \# local misassemblies                     & 8           & 11        \\[1.0em]
        Misassembled contigs length                & 0           & 32563     \\[1.0em]
        Genome fraction (\%)                       & 93.880      & 93.966    \\[1.0em]
        \# mismatches per 100 kbp                  & 3.58        & 4.70      \\[1.0em]
        \# indels per 100 kbp                      & 6.21        & 5.88      \\\bottomrule
      \end{tabularx}
      \end{table}

      \vspace{1em}
      {\small
        Command-line parameters used for assembly: \linebreak
        \texttt{--only-assembler -k 21,33,55,77,99}
      }

    \end{block}
    \begin{block}{Acknowledgements}
        This work was supported by the Government of the Russian Federation
        (grant 11.G34.31.0018) and RFBR (grant 12-01-00747-a).

        We would like to thank Sergey Nikolenko and Alla Lapidus for their
        insightful comments and fruitful discussions throughout the project.
    \end{block}
    \begin{block}{Further information}
      \vspace{1em}
      \color{red}{Put here a link to SPAdes}
      \vspace{2.7em}
    \end{block}
  \end{textblock}


  \begin{textblock}{0.215}(0.01, 0.64)

    \begin{detailsblock}{Pairwise distance calculation \phantom{p}}
      We use 5-base lookahead to compute distance between $k$-mers in homopolymer-space. Helper
      table stores precomputed values of
      \begin{eqnarray*}
        H_k: \mathscr{N}^{k} \times \mathscr{N}^{k} \to \left\{\mathrm{Insertion, Deletion,
          Mismatch}\right\}, \ k=1,2,3,4,5\,\mathrm{.}
      \end{eqnarray*}
      The chosen value of 5 is~a~trade-off between accuracy and speed.
      
      (Mapped reads from B7-295.bam dataset were used for training.)

      \begin{center}
        \begin{table}[h!]
          \begin{tabular}{llllllllllllllll}
            G  &  \cellcolor{blue!30}A  &  \cellcolor{blue!30}G  &  \cellcolor{blue!30}T  &  \cellcolor{blue!30}A  &  \cellcolor{blue!30}C  &  A  &  \cellcolor{green!30}C  &  \cellcolor{green!30}T  &  \cellcolor{green!30}G  &  \cellcolor{green!30}T  &  \cellcolor{green!30}C  &  G  &  \cellcolor{red!30}T  &  \cellcolor{red!20!orange!30}C  &  \cellcolor{red!20!orange!30}G  \\
            G  &  \cellcolor{blue!30}T  &  \cellcolor{blue!30}G  &  \cellcolor{blue!30}T  &  \cellcolor{blue!30}A  &  \cellcolor{blue!30}C  &  A  &  \cellcolor{green!30}T  &  \cellcolor{green!30}G  &  \cellcolor{green!30}T  &  \cellcolor{green!30}C  &  \cellcolor{green!30}G  &  \cellcolor{red!30}A  &  \cellcolor{red!20!orange!30}T  &  \cellcolor{red!20!orange!30}G  &  C  \\
          \end{tabular}
        \end{table}
        \begin{align*}
          H_5(\mathrm{AGTAC}, \mathrm{TGTAC}) &= \mathrm{Mismatch} \\
          H_5(\mathrm{CTGTC}, \mathrm{TGTCG}) &= \mathrm{Deletion} \\
          H_3(\mathrm{TCG}, \mathrm{ATG}) &= \mathrm{Mismatch} \\
          H_2(\mathrm{CG}, \mathrm{TG}) &= \mathrm{Mismatch}
        \end{align*}
      \end{center}

      \vspace{-.3em}
    {\large Algorithm}
    \begin{algorithmic}
      \State \textbf{Input:} $x, y \in \mathscr{N}^+$ --- homopolymer-space $k$-mers in nucleotide alphabet
      \State \textbf{Output:} $dist$ --- approximate distance between $x$ and $y$
      \vspace{0.3cm}
      \State $pos.x \gets 0$; $pos.y \gets 0$; $dist \gets 0$;
      \While{$pos.x < |x|\textrm{ and }pos.y < |y|$}
      \If{$x[pos.x] = y[pos.y]$}
      \State $pos.x \gets pos.x + 1$; $pos.y \gets pos.y + 1$;
      \Else
      \State $k \gets \mathrm{min}(5, |x| - pos.x, |y| - pos.y)$;
      \State adjust $pos.x$ and $pos.y$ according to
      \State \hspace{1cm} $H_k(x[pos.x\textrm{ .. }pos.x + k - 1], y[pos.y\textrm{ .. }pos.y + k - 1])$;
      \State $dist \gets dist + 1$;
      \EndIf
      \EndWhile
    \end{algorithmic}
    \end{detailsblock}
  \end{textblock}

  \begin{textblock}{0.23}(0.230, 0.64)
    \begin{detailsblock}{Homopolymer-space $k$-mer clustering}
      The starting point is single-linkage clustering of homopolymer-space $k$-mers. Two homopolymer-space
      $k$-mers belong to the same cluster if~the~distance between their nucleotide representations does not exceed one.

      We reduce quadratic time requirements of the naive algorithm by noticing that if distance between two $k$-mers is less or
      equal to one, they share a common substring of length at least $\lfloor k/2 \rfloor$. This allows us to group
      $k$-mers into smaller blocks sharing a substring, and then use the quadratic algorithm for each block.
      
      In order to detect all pairs of connected $k$-mers with such grouping, ranges $(0\textrm{ .. }\lfloor k/2 \rfloor
      - 1)$, $(1\textrm{ .. }\lfloor k/2 \rfloor)$, $\dots$, $(\lfloor (k + 1)/2 \rfloor\textrm{ .. }k - 1)$ are to be examined.
      Despite of this, we use only the first and the last of the ranges, to speed up the clustering step. 
      This increases the number of \emph{singletons} --- clusters consisting of only one $k$-mer, usually erroneous ---
      but has little impact on correction performance because of adjacent read $k$-mers making much larger contribution
      into the consensus scores (the number of singleton occurrences is usually low).
     
      \vspace{0.3em} 
      \emph{partition}($\mathscr{K}$, $i$, $j$) $=$
      $\left\{B_s : \bigcup_{s\in \mathscr{H}^{j - i + 1}}B_s = \mathscr{K},\;\forall x\in B_s\ x[i\textrm{ .. }j] =
        s\right\}$

      {\large Algorithm}

      \begin{algorithmic}
        \State \emph{kmers} $\gets$ homopolymer-space $k$-mers seen in the data
        \State \emph{components} $\gets$ $\{\{k\} : k \in \textrm{\emph{kmers}}\}$
        \State \emph{blocksL} $\gets$ \emph{partition}(\emph{kmers}, $0$, $\lfloor k/2 \rfloor - 1$)
        \State \emph{blocksR} $\gets$ \emph{partition}(\emph{kmers}, $\lfloor (k+1)/2 \rfloor$, $k - 1$)
        \For{\textbf{each} \emph{block} \textbf{in} \emph{blocksL}, \emph{blocksR}}
          \For{\textbf{each} $i \in \textrm{\emph{block}}$}
            \For{\textbf{each} $j \in \textrm{\emph{block}}$}
              \If{$\mathrm{dist}(i, j) \leq 1$}
                 \State join the components to which $i$ and $j$ belong
              \EndIf
             \EndFor
           \EndFor
         \EndFor
      \end{algorithmic}
    \end{detailsblock}
  \end{textblock}

    \begin{textblock}{0.200}(0.465, 0.64)
    \begin{detailsblock}{Subclustering}
      A cluster obtained from the initial process may contain $m \geq 2$ homopolymer-space $k$-mers from the genome.
      In this case, we split the cluster into $m$ subclusters by running $k$-means algorithm on it.

      The subtle question is how to determine $m$. Currently we just set it to be the number of the
      cluster elements with quality within machine epsilon of $1$, where \emph{quality} of $x \in \mathscr{H}^k$
      is defined as $\Pr(x\textrm{ is genomic})$.
      For the set of reads $\mathscr{R} \subset \mathscr{N}^+$ the quality of $x \in \mathscr{H}^k$ is computed as
      \begin{equation*}
        1 - \prod_{r \in \mathscr{R}}\prod_{\substack{0 \leq m \leq |r| - |x'|, \\ r[m\textrm{ .. }m + |x'| - 1] =
            x'}}\left(1 - \prod_{m\leq n < m + |x'|}\Pr(r[n]\textrm{ is correct})\right)\,\mathrm{,}
      \end{equation*}
      where $x' \in \mathscr{N}^+$ is the sequence of nucleotides in $x$. 
    \end{detailsblock}

    \begin{detailsblock}{Cluster size distribution}
      \begin{figure}
        \includegraphics[width=0.97\textwidth]{images/clustersize}
      \end{figure}
    \end{detailsblock}
  \end{textblock}

  \begin{textblock}{0.115}(0.670, 0.64)
    \begin{detailsblock}{Typical cluster}
      \begin{table}[ht!]
        \scalebox{0.666}{
          \ttfamily
          \begin{tabular}{l|r|r}
            \textbf{homopolymer-space 16-mer}  & \textbf{\#}                             & \textbf{qual.}                          \\ \hline 
            \ CCCGTTTGT-GCGCCC-GG--CATTT-GAT   & \cellcolor{green!100!red!20!white} 562 & \cellcolor{green!100!red!20!white} 1.00 \\
            \ CCCGTTTGT-GCGCCC-GG--CATT--GAT   & \cellcolor{green!2!red!20!white} 13    & \cellcolor{green!100!red!20!white} 1.00 \\
            \ CCCGTT-GT-GCGCCC-GG--CATTT-GAT   & \cellcolor{green!2!red!20!white} 10    & \cellcolor{green!100!red!20!white} 1.00 \\
            \ CCCGTTTGT-GCGCCCCGG--CATTT-GAT   & \cellcolor{green!1!red!20!white} 6     & \cellcolor{green!100!red!20!white} 1.00 \\
            \ CCCGTTTGT-GCGCCC-GG--CATTTTGAT   & \cellcolor{green!1!red!20!white} 5     & \cellcolor{green!99!red!20!white} 0.99  \\
            \ \ CCGTTTGT-GCGCCC-GG--CATTT-GAT  & \cellcolor{green!1!red!20!white} 4     & \cellcolor{green!100!red!20!white} 1.00 \\
            CCCCGTTTGT-GCGCCC-GG--CATTT-GAT    & \cellcolor{green!1!red!20!white} 4     & \cellcolor{green!98!red!20!white} 0.98  \\
            \ \ \ CGTTTGT-GCGCCC-GG--CATTT-GAT & \cellcolor{green!1!red!20!white} 3     & \cellcolor{green!100!red!20!white} 1.00 \\
            \ CCCGTTTGT-GCGCC--GG--CATTT-GAT   & \cellcolor{green!1!red!20!white} 3     & \cellcolor{green!98!red!20!white} 0.98  \\
            \ CCCGTTTGT-GCGCCC-GG--CATTT-GAAT  & \cellcolor{green!1!red!20!white} 3     & \cellcolor{green!98!red!20!white} 0.98  \\
            \ CCCGTTTGT-GCGCCC-GGA-CATTT-GA    & \cellcolor{green!1!red!20!white} 3     & \cellcolor{green!98!red!20!white} 0.98  \\
            \ CCCGTTTGT-GCGCCC-GG--C-TTT-GATG  & \cellcolor{green!1!red!20!white} 3     & \cellcolor{green!97!red!20!white} 0.97  \\
            \ CCCGTT-GTTGCGCCC-GG--CATTT-GAT   & \cellcolor{green!0!red!20!white} 2     & \cellcolor{green!94!red!20!white} 0.94  \\
            \ CCCGTTTGTTGCGCCC-GG--CATTT-GAT   & \cellcolor{green!0!red!20!white} 2     & \cellcolor{green!94!red!20!white} 0.94  \\
            \ CCCGTTTGGTGCGCCC-GG--CATTT-GAT   & \cellcolor{green!0!red!20!white} 2     & \cellcolor{green!93!red!20!white} 0.93  \\
            \ CCCGTTTGT-GCGCCC-GG--CATTT-GAG   & \cellcolor{green!0!red!20!white} 2     & \cellcolor{green!89!red!20!white} 0.89  \\
            \ CCCGTTTGT-GCGCCC-GG--CATTTGGAT   & \cellcolor{green!0!red!20!white} 2     & \cellcolor{green!74!red!20!white} 0.74  \\
            \ CCCGTTTGT-GCGCTC-GG--CATTT-G     & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!97!red!20!white} 0.97  \\
            \ CCCGTTTGTG-CGCCC-GG--CACTTTGA    & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!87!red!20!white} 0.87  \\
            \ CCCGTTTGTG-CGCCC-GG-CCATTT-GAT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!85!red!20!white} 0.85  \\
            \ CCCGTTTGTGC-GCCC-G---CATTT-GAT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!83!red!20!white} 0.83  \\
            \ CCCGTTTGTGC-GCCC-GG--CATTT-GGTG  & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!74!red!20!white} 0.74  \\
            \ \ CCGTTTGTGCCGCCC-GG--CATTT-GAT  & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!71!red!20!white} 0.71  \\
            \ CCCGTTTTGTGCGCCC-GG--CATTT-GAT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!69!red!20!white} 0.69  \\
            \ CCCGTTTGT-GCGCCC-GGTACATTT-G     & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!68!red!20!white} 0.68  \\
            \ CCCGTTTGT-GCGCCC-GG--CATTTCGA    & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!68!red!20!white} 0.68  \\
            \ CCCGTTTGTGCCGCCC-GG--CATTT-GAT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!66!red!20!white} 0.66  \\
            \ \ CCGTT-GT-GCGCC--GG--CATTT-GAT  & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!66!red!20!white} 0.66  \\
            \ CCCGTTTGTG-CGCCC-GGG-CATTT-GAT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!61!red!20!white} 0.61  \\
            \ CCCGTTTGT-GCGCCC-GGACCATTT-GA    & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!61!red!20!white} 0.61  \\
            \ CCCGTTTGTGC-GCCC-GG--CATTTTGTA   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!58!red!20!white} 0.58  \\
            \ CCCGTTTGTGGCGCCC-GG--CATTT-GAT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!57!red!20!white} 0.57  \\
            \ CCCGTTTGTGC-GCCC-GGA-CATTT-AG    & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!56!red!20!white} 0.56  \\
            \ CCCGTTTGT-GCGCCC-GG--CATTT-AGT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!55!red!20!white} 0.55  \\
            \ \ CCGTTTGTGC-GCC--GG--CATTT-GAT  & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!54!red!20!white} 0.54  \\
            \ CCCGTT-GTGC-GCCC-GGG-CATTT-GAT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!44!red!20!white} 0.44  \\
            \ CCCGTTTGTGCCGCCCTGG--CATTT-GA    & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!43!red!20!white} 0.43  \\
            \ CCCGTTTTGTGCGCCC-GG--CATTTTGAT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!36!red!20!white} 0.36  \\
            \ CCCGTTTGGTGCGCCCCGG--CATTT-GAT   & \cellcolor{green!0!red!20!white} 1     & \cellcolor{green!22!red!20!white} 0.22  \\
            \end{tabular}
        }
      \end{table}
    \end{detailsblock}
  \end{textblock}
  \begin{textblock}{0.205}(0.790, 0.64)
    \begin{detailsblock}{Error correction algorithm}
      \vspace{0.7em}
      \begin{algorithmic}
        \For{\textbf{each} \textit{read} from the dataset}
      \item \hspace{\algorithmicindent}\emph{(Producing contiguous corrected parts of \emph{read})}
          \For{\textbf{each} homopolymer-space \textit{kmer} from the read}
            \State \textit{center} $\gets$ center of the cluster to which \textit{kmer} belongs;
            \If{\textit{center} quality is more than user-specified threshold \textbf{and}\\
              \hspace{4.0em}\textit{center} bases agree with the previous ``good'' center}
              \State include \textit{center} into consensus score calculation;
            \Else
              \State yield new corrected part from current consensus;
              \State trim homopolymer runs with low consensus score from ends;
              \State reset consensus table and start a new part;
            \EndIf
          \EndFor
      \item \hspace{\algorithmicindent}\emph{(Combining corrected parts)}
          \While{there are two or more parts}
            \State \emph{curr} $\gets$ first part; \emph{next} $\gets$ second part;
            \State align last 8 homopolymer runs of \emph{curr} against the read;
            \State align first 8 homopolymer runs of \emph{next} against the read;
            \If{there is a gap on the read between the two parts}
              \State copy read homopolymer runs as is;
            \Else
              \State select homopolymer runs with higher consensus score\\
          \hspace{6.0em}from the intersection of the two parts;
            \EndIf
            \State replace \emph{curr} and \emph{next} with the combined part;
          \EndWhile
      \item \hspace{\algorithmicindent}\emph{(Optionally, attaching uncorrected end)}
           \State align last 8 runs of the last chunk against the read sequence;
           \State append read homopolymer runs after the last aligned run.
        \EndFor
      \end{algorithmic}
      \vspace{0.6em}
  \end{detailsblock}
\end{textblock}

\end{frame}
\end{document}
